### PR TITLE
Autoscaling

### PR DESCRIPTION
**Pull request by jdoliner**
This adds a prototype of dynamic parallelism functionality:

- ~~Pipelines don't spin up if no datums will run due to failed inputs.~~
- Pipelines still spin up if someone does an empty commit or if all datums will be skipped.
- Pipelines spin up a single a single worker to start, when it sees unclaimed tasks that could be claimed by new workers it spins up more workers. Up to the parallelism of the pipeline.

### CODE REVIEW

**Comment by jdoliner** [comment]
Things that we need to think about before we can merge this:

- How data is passed around. I added two fields to `PipelineInfo` for this PR. It's unclear if these fields make sense there or should be somewhere else.
- Polling, currently this polls the pipeline to see how much unclaimed work there is and scales up if there's enough. We generally don't use polling throughout the codebase. I think it would be pretty easy to turn this in to a watch. I do think polling is not as bad here as in other places since a lot of the edge cases don't matter. I.e. one worry with polling is that you might not notice the unclaimed work in time for the scale up to happen, but that just means the job finished before we could scale up in which case we actually don't want to scale up at all.
- PipelineController control flow, this code works by allowing monitor pipeline to scale up and down the worker RC, generally all rc modifications happen within the control loop and these changes do retrigger the control loop because they modify the rc in k8s. I'm not sure if this is a problem, it seems to work fine but I defer to original authors of pipeline controller.
- Currently the pipeline won't spin up if the commits can't be processed due to failed inputs. Presently it doesn't create a job or close the commits, closing the commits is easy enough. I'm a little nervous about creating the job from within the pps master rather than the worker master. It seems like a potential race condition. Probably not too bad but something that needs to be thought about.

File: src/server/worker/pipeline/transform/registry.go
Comment (Tryneus) [inline src/server/worker/pipeline/transform/registry.go:398]: I'm not sure this change is doing what you expect.  Imagine a pipeline with `concurrency=20` and `maxDatumsPerTask=3`.  If I've read this correctly, we have the following behavior before vs after:

| datums | tasks (before) | tasks (after) |
| -------- | --------------- | ------------- |
| 10 | 10 | 10 |
| 20 | 20 | 20 |
| 40 | 20 | ***80*** |
| 80 | 26 | 26 |
| 160 | 53 | 53 |

Note that with increasing datum numbers, the code before has a smooth transition in number of tasks with increasing number of datums - the new code has a discontinuity in the `else` case where the number of tasks spikes, then goes back down in the `else if` case.

**Review by Tryneus — COMMENTED** [review]
I think to productionize this, we will also need the capability of scaling down a running pipeline when there are not enough tasks to saturate all the workers.  If the PPS master could tell the number of idle workers, this would probably be pretty easy to do.  Perhaps having the workers mark themselves as idle in etcd if they haven't had a task to run for `n` seconds would do?

**Comment by jdoliner** [comment]
So the issue with scale down is that there isn't a straightforward way to tell k8s which pods you want it to kill when you scale down. So when you scale down it might kill the one worker that's still doing something which would lose all the work. There's some kind of crazy stuff you can do with labels to work around this but it seemed like to much to do for the first pass on this and it may just be too hacky to do at all.

**Comment by Tryneus** [comment]
It seems like we could do a scale down if we moved to `StatefulSet`s ([ref](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#deployment-and-scaling-guarantees)) for workers, but even then it only kills pods in reverse order of creation, if I understand it correctly.  There are some tradeoffs here, and we'd need an additional change to make sure work tends to go to earlier worker pods, but that sounds doable.

**Comment by jdoliner** [comment]
I think it's going to be pretty tricky to have work tend to go to early pods. In general we want to spin up a pod only when there's work for it to do, but we have no way of knowing how long that work is going to take. It could happen that the work we assign to the 10th pod happens to take the longest and then we're stuck with pods 1-9 doing nothing while it finishes. A real solution to this is probably going to be a custom controller at which point we should have direct control over which pods get spun up and spun down.

File: src/server/pkg/work/work.go
Comment (jdoliner) [reply src/server/pkg/work/work.go:409]: This is now done transactionally which should make it impossible to get a negative number. But in case we do it'll return 0.

**Comment by jdoliner** [comment]
So after thinking about it a little more I'm not sure that scale down is going to work well, even if it was easy to implement. The thing I realized is that even if there's only one worker left processing, after that datum is done we need to do a merge and that process is distributed. If the processing is really going to take a long time it might make sense to spin down the idle workers and then spin them up for the merge. But we have no way of knowing how long it's going to take so we may well just be wasting time scheduling and descheduling pods. I think a better approach is a work stealing system whereby idle workers can "steal" datums from other workers to help finish the remaining chunks more quickly. Of course if it's just a single, long-running, datum then it can't be stolen and we're stuck just waiting for it to finish.

**Review by msteffen — COMMENTED** [review]
Two quick low-level questions

File: src/server/pachyderm_test.go
Comment (jdoliner) [reply src/server/pachyderm_test.go:259]: Oh I changed this to set `standby: true` because that was how I was testing it for a while. I'll change it back because I'm writing a whole new test for this and there's no reason to change this one.

Comment (jdoliner) [reply src/server/pachyderm_test.go]: Changed back now.

File: src/client/pps/pps.proto
Comment (jdoliner) [reply src/client/pps/pps.proto:467]: Added.

File: src/server/worker/pipeline/transform/registry.go
Comment (jdoliner) [reply src/server/worker/pipeline/transform/registry.go:398]: @Tryneus alright I thiiink I fixed this. It's a little hard to reason about but I think the issue was basically due to updating the default concurrency but not the `numDatums/maxDatumsPerTask > reg.concurrency` part of the conditional I was increasing the number of datums in some cases that I didn't mean to. I just change the condition to `numDatums/maxDatumsPerTask > reg.concurrency*taskGranularity` which is the equivalent of what it was before. Lmk if this pass the Marc test though.

File: src/server/pps/server/monitor.go
Comment (msteffen) [inline src/server/pps/server/monitor.go:362]: I think the main risk here is that this code could race with the pipeline controller, but even if it does get the RC into a bad state (e.g. scaling up when a pipeline is supposed to be in standby) the pipeline poller will eventually come along and fix things (fixing weird states due to races is basically the purpose of the pipeline poller).

I think this would all ideally happen in the pipeline controller if possible (to avoid having two concurrent goroutines modifying the RC. It might make sense to add a TODO to that effect?), but I don't think the code will break or anything if we postpone that (and even if we did make that change and all writes came from the same goro, there's no way to completely prevent racing writes to the RC without a write fence). We might just have the occasional pipeline take a long time to go into standby or something.

File: src/server/pps/server/api_server.go
Comment (msteffen) [inline src/server/pps/server/api_server.go:2972]: Just curious, are there other places where we instantiate a Worker instance in pachd? I'm trying to think of weird states that this struct might get into—probably it's fine to leave it, but part of me wonders if it's worth handling e.g. `Not Found` in case this races with `UpdatePipeline`? (is `NotFound` even the error you'd get? Maybe it makes sense to just leave this alond too)

**Review by msteffen — APPROVED** [review]
A few questions but overall LGTM

File: src/server/worker/datum/iterator.go
Comment (jdoliner) [reply src/server/worker/datum/iterator.go:81]: This was one of the most hated features of the existing system because it intentionally grouped all of the largest datums into the same chunk which frequently led to stragglers. It wasn't strictly necessary to make the rest of autoscaling work, but it seemed like a good opportunity to fix it (honestly it was way overdue) and might help with people's overall perception of the autoscaling changes.

File: src/server/pps/server/monitor.go
Comment (jdoliner) [reply src/server/pps/server/monitor.go:362]: Yeah I had a similar worry here. Figuring out how to propagate this info back to the controller seemed a little tricky. I think I agree that the worst thing I could see happening here is a pipeline being slow to go into standby, I'll add the TODO though.

File: src/server/pps/server/api_server.go
Comment (jdoliner) [reply src/server/pps/server/api_server.go:2972]: So I don't think you can get `not found` because `UnclaimedTasks` is implemented as taking the difference between two count operations in etcd and I don't think those can ever return `not found` they just return 0 if they don't find anything.

File: src/server/pachyderm_test.go
Comment (jdoliner) [inline src/server/pachyderm_test.go:5964]: This change was necessary due to the change to the ordering in datum iterators.

File: src/server/worker/pipeline/transform/registry.go
Comment (Tryneus) [reply src/server/worker/pipeline/transform/registry.go:398]: yes, this would do it

**Review by Tryneus — APPROVED** [review]
LGTM

**Pull request merged**
